Feature: xYield ERC-4626

evchip opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
- create xYield contract, inheriting ERC-4626
  - add `deposit` and `depositFrom` methods for same chain and remote chain deposits, respectively
  - add method `updateTotals` with permission check for guardian (offchain program running in TEE) in order to update global shares and user positions
  - add placeholders for unimplemented method `rebalance`
- create scripts to deploy xYield contracts to Arbitrum and Base mainnets
- add deployment metadata for xYield v0.0.1 deployment
<!--- Describe your changes in detail -->

## Motivation and Context
Checkpoint with boilerplate functionality in order to deploy and unblock offchain work streams.
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
Fork tests and unit tests for `deposit` and `depositFrom` methods
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes (remove all unchecked types)

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
-   [x] New feature (non-breaking change which adds functionality)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

-   [x] If my change requires a change to the documentation, I have updated the documentation accordingly, and in either case, have checked this box to attest to my assessment of this requirement with regard to my change.
-   [x] If my change requires additions or updates to any deployment scripts to ensure that the protocol is functional (Makefile, Dockerfile, Forge Script, etc.), I have made these changes, and in either case, have checked this box to attest to my assessment of this requirement with regard to my change.
-   [x] If my change requires additional test coverage, I have created those tests accordingly, and in either case, have checked this box to attest to my assessment of this requirement with regard to my change.

evchip committed:
feat(xYield): initial commit of xYieldVault

establishing simple deposits and withdrawals; add unimplemented functions as todos

evchip committed:
feat(forge modules): install euler vault kit as dependency

evchip committed:
feat(remappings.txt): add euler path and a fix for

bad dependency path on ethereum vault connector

evchip committed:
chore(xYield): forge fmt

evchip committed:
fix(xYieldVault): fix incorrect call in totalAssets

instead of calling yieldProtocol.totalAssets we need to call yieldProtocol.convertToAssets

evchip committed:
test(xYieldForkTest): create 3 tests:

deposit, withdraw, and interest accrual

evchip committed:
feat(xYield): wIP

evchip committed:
fix(xYieldVault): rm T1ERC7683

evchip committed:
feat(xYield): wIP checkpoint

evchip committed:
feat(xYield): wIP checkpoint

evchip committed:
refactor(xYield): wIP cleanup

evchip committed:
refactor(xYield): wip cleanup

evchip committed:
fix(foundry.toml): change solc version back to 0.8.28

to prevent compile issues with incompatible solc versions

evchip committed:
feat(DeployxYieldVault): create scripts to deploy xYield to Arb and Base

evchip committed:
feat(contracts/deployments/xyield/v0.0.1.json): add deployment metadata

evchip committed:
chore(contracts): forge fmt

evchip committed:
chore(workflows/contracts.yml): add env var ARBITRUM_RPC for fork tests

evchip committed:
chore(contracts.yml): add ARBITRUM_BLOCK secret

aharvet commented on contracts/src/xYield/xYieldTest.t.sol:
```diff
@@ -0,0 +1,172 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.25;
+
```
Maybe we should put those test files in the test folder ?

evchip replied:
fixed, thanks.

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,288 @@
+    uint256 public minRebalanceGap;
+    uint256 public lastRebalanceTime;
+
+    event DepositRemote(address indexed sender, address indexed owner, uint256 assets, uint256 shares);
```
Let's group this event with the other ones ?

evchip replied:
fixed, thanks.

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,288 @@
+            underlying.approve(_yieldProtocol, type(uint256).max);
+        }
+
+        _transferOwnership(msg.sender);
```
I think ownership is already transfered to sender in Ownable isn't it ?

evchip replied:
This is Ownable2Step. The first step is `_transferOwnership`, the second step is for the recipient to call `acceptOwnership`

aharvet replied:
I mean Ownable2Step inherits Ownable, and `_transferOwnership(_msgSender())` is already called in Ownable constructor

evchip replied:
Got it, removed

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,288 @@
+    }
+
+    function deposit(
+        uint256 assets,
```
assets is not very descriptive to me here, I would rather go for amount

evchip replied:
fixed, thanks.

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,288 @@
+        uint256 shares = previewDeposit(assets);
+        _deposit(msg.sender, receiver, assets, shares);
+
+        return shares;
```
How could this return value be used ?

evchip replied:
If another contract (e.g. a vault or smart account) calls deposit, they get back the number of shares minted to them from their deposit.

aharvet replied:
Ok makes sense

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,288 @@
+        public
+        virtual
+        override
+        nonReentrant
```
Do we actually need to use nonReentrant ?
There is no ether transfer which is the biggest reentrancy risk factor, the interaction with external contracts are at the very end which respects check effect interaction pattern and those are white listed.

evchip replied:
good point, fixed

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,288 @@
+    // global shares is updated
+    function _depositFrom(address caller, address receiver, uint256 assets, uint256 shares) internal {
+        IERC20(asset()).transferFrom(caller, address(this), assets);
+        yieldProtocol.deposit(assets, address(this));
```
We don't account for total assets and supply here because it's already done on initial deposit chain right ?

evchip replied:
total assets and supply will be updated in subsequent txs `updateTotals` called by the TEE guardian. The `virtualTotalAssets` and `virtualTotalSupply` will be updated with the same values across all chains. The user who deposited from the remote chain will be minted his share tokens there. This prevents a scenario where users who already have shares in the pool are able to withdraw the remote deposit user's assets.

aharvet replied:
Ok makes sense

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,288 @@
+    }
+
+    // called on behalf of a user who has deposited from a remote chain
+    function depositFrom(uint256 assets, address receiver) public nonReentrant whenNotPaused returns (uint256 shares) {
```
Shouldn't we restrein access to our infra only ? to avoid anyone to mess up the crosschain accounting
Or maybe use something to prove an previous deposit has been done on another chain ?

evchip replied:
This will be called by solvers, I believe. The deposit is actually being done on this chain, it's simply an affordance to users who want their share tokens minted on another chain. I don't think we need to restrict it. After the solver deposits, we will pick up the event offchain and mint the shares for them on the remote chain.

But I see that I should add the chain ID as a param so we know which chain to mint to, and also to check whether it is a supported chain.

aharvet replied:
Ok makes sense, I think I'm just not 100% aware of the backend role yet

aharvet replied:
@evchip just to make sure, so if a user would call this one directly (and not a solver) he would just receive the shares on a different chain right ?

evchip replied:
Yes, exactly

aharvet reviewed (commented):
NOT FINISHED

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,288 @@
+        // TODO - update incrementally like virtualTotalAssets using proof of remote supply change
+        virtualTotalSupply = totalSupply_;
+        _updateBalances(_recipients, _amounts);
+        virtualTotalAssets = this.totalAssets();
```
Is there a reason why we need to use this. here and below instead of accessing the variable directly ?

evchip replied:
good catch, fixed

evchip committed:
refactor(xYieldTest.t.sol): move to tests dir

evchip committed:
refactor(xYieldForkTest): move to test dir

evchip committed:
refactor(xYield): group deposit with other events

evchip committed:
refactor(MockYieldProtocol): move to test dir

evchip committed:
refactor(xYieldVault.sol): change assets to amount

evchip committed:
refactor(xYield tests): fix import paths

evchip committed:
refactor(xYieldVault): prepend underscore to method params for clarity

evchip committed:
refactor(xYieldVault): remove unneeded reentract protection on deposit

evchip committed:
refactor(xYieldVault): remove this from totalAssets

evchip committed:
chore(contracts): merge upstream canary

evchip committed:
fix(xYieldVault): fix incorrect var naming

evchip committed:
feat(xYieldVault): add chainId param to depositFrom

in order to track remote chain

evchip committed:
chore(xYieldVault): forge fmt

evchip committed:
refactor(xYieldVault): rm _transferOwnership -- redundant

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+    error NotGuardian();
+    error ZeroAmount();
+    error InvalidChain();
+    error RebalanceTooFrequent();
```
Error not used. Maybe it's for future ?

evchip replied:
removed

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+    IERC4626 public yieldProtocol;
+
+    uint256 public virtualTotalAssets;
+    uint256 public virtualTotalSupply;
```
Not necessarily an AI but I have a hard time wrapping my head around those variable names after multiple read of the contract. Maybe `virtualTotalDeposit` and `virtualTotalLiquidity` could be more explicit ?

evchip replied:
I'm just following the 4626 convention but prefixing to `virtual` to make clear it's not necessarily the same as `totalAssets` and `totalSupply`. Let me do a write up on this so it's more clear.

aharvet replied:
I'm fine with virtual, it's rather the assets and supply terms. But no worries I just wanted to mention it in case you feel the same way.

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+        virtualTotalSupply += _shares;
+        _mint(_receiver, _shares);
+
+        IERC20(asset()).transferFrom(_caller, address(this), _amount);
```
We should use `safeERC20` for transfers.

evchip replied:
done

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+        address _receiver,
+        uint64 _chainId
+    )
+        public
```
I think we can set this one as external ?

evchip replied:
done

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+        public
+        virtual
+        override
+        nonReentrant
```
I think my previous comment about nonReentrant applies to all of them

evchip replied:
removed

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+        virtualTotalAssets -= _amount;
+        virtualTotalSupply -= _shares;
+        _burn(_owner, _shares);
+        IERC20(asset()).transfer(_receiver, _amount);
```
We should use safeERC20 for transfers.

evchip replied:
done

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+    )
+        internal
+    {
+        IERC20(asset()).transferFrom(_caller, address(this), _amount);
```
We should use safeERC20 for transfers.

evchip replied:
done

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+import { ReentrancyGuard } from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import { Pausable } from "@openzeppelin/contracts/security/Pausable.sol";
+
+contract xYieldVault is ERC4626, Ownable2Step, ReentrancyGuard, Pausable {
```
Should we have an interface for it ?

evchip replied:
done

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+    // TODO - submit proofs for remote deposits
+    function _updateBalances(address[] calldata _recipients, uint256[] calldata _amounts) private {
+        if (_recipients.length != _amounts.length) revert LengthMismatch();
+        for (uint256 i = 0; i < _recipients.length; i++) {
```
I am afraid this won't scale, f we get a lot of users this will be very expensive.
What do you think ?

evchip replied:
it's a good point. there are two ways that we could get away with users minting share tokens:
1. mint them when users initiate a remote deposit request. if the intent fails, we burn it for them. intuitively this seems a bit risky because we (temporarily) have share tokens minted before the underlying asset has reached the vault. maybe it's fine as long as we have a way to make it atomic.
2. we post a merkle root to each chain which proves their deposit landed, and expose proofs that the user can submit which if valid allow them to mint share tokens.

let me know which approach you think is better, or if you have a better idea.

aharvet replied:
I thought of 2 too. I agree that 1 is risky.
I just don't know if introducing this extra step for users resonate with the whole point of the product.

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+    mapping(uint64 => address) public siblingVaults;
+
+    uint256 public minRebalanceGap;
+    uint256 public lastRebalanceTime;
```
Is it meant that we don't use these variables yet ?

evchip replied:
removed

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.25;
```
Version should be fixed

evchip replied:
done

aharvet commented on contracts/src/test/xYield/xYieldTest.t.sol:
```diff
@@ -0,0 +1,186 @@
+        assertEq(vault.totalSupply(), shares, "total supply");
+        assertEq(vault.totalAssets(), depositAmount, "total assets");
+        assertEq(usdc.balanceOf(user1), INITIAL_DEPOSIT - depositAmount, "usdc balance of user1");
+    }
```
Maybe we should also test other functions like depositFrom, mint, redeem... at least for non regression ?

evchip replied:
`depositFrom` is covered in the fork test. I can do mint and redeem, but not a high priority IMO. I'll make a TODO inline.

aharvet replied:
Sure

aharvet reviewed (commented):
A few more comments

aharvet commented on contracts/src/xYield/xYieldVault.sol:
```diff
@@ -0,0 +1,292 @@
+    uint256 public virtualTotalSupply;
+    bool public isActiveChain;
+
+    mapping(uint64 => address) public siblingVaults;
```
Is it relevant to use uint64 here as there's no gas saving ?

evchip replied:
most ethereum clients use uint64 for chainId. i'm just sticking to that convention

aharvet replied:
Okay makes sense

evchip committed:
fix(xYieldVault): fix incorrect formula for totalAssets

total assets should be based on xYield's balance of share tokens; change erc20 transfer calls to
safeTransfer

evchip committed:
test(xYieldForkTest.t.sol): add TODO for mint and redeem tests

evchip committed:
chore(IxYieldVault): add interface for xYield vault

evchip committed:
refactor(xYieldVault): rm unused vars and methods for rebalancing

evchip committed:
chore(contracts): forge fmt

aharvet reviewed (approved):
LGTM. We just need to keep in mind the loop in _updateBalances topic for next iterations.
